cpp/src/parquet/column_writer.cc

```diff
@@ -645,23 +686,64 @@ class TypedColumnWriterImpl : public ColumnWriterImpl, public TypedColumnWriter<
 
     if (properties->statistics_enabled(descr_->path()) &&
         (SortOrder::UNKNOWN != descr_->sort_order())) {
-      page_statistics_ = TypedStats::Make(descr_, allocator_);
-      chunk_statistics_ = TypedStats::Make(descr_, allocator_);
+      page_statistics_ = MakeStatistics<DType>(descr_, allocator_);
+      chunk_statistics_ = MakeStatistics<DType>(descr_, allocator_);
     }
   }
 
   int64_t Close() override { return ColumnWriterImpl::Close(); }
 
   void WriteBatch(int64_t num_values, const int16_t* def_levels,
-                  const int16_t* rep_levels, const T* values) override;
+                  const int16_t* rep_levels, const T* values) override {
+    // We check for DataPage limits only after we have inserted the values. If a user
+    // writes a large number of values, the DataPage size can be much above the limit.
+    // The purpose of this chunking is to bound this. Even if a user writes large number
+    // of values, the chunking will ensure the AddDataPage() is called at a reasonable
+    // pagesize limit
+    int64_t value_offset = 0;
+    auto WriteChunk = [&](int64_t offset, int64_t batch_size) {
+      int64_t values_to_write =
+          WriteLevels(batch_size, def_levels + offset, rep_levels + offset);
+      // PARQUET-780
+      if (values_to_write > 0) {
+        DCHECK_NE(nullptr, values);
+      }
+      WriteValues(values + value_offset, values_to_write, batch_size - values_to_write);
+      CommitWriteAndCheckPageLimit(batch_size, values_to_write);
+      CheckDictionarySizeLimit();
```

Comment:
We can't have it there because the data page limit logic is needed in the direct dictionary write path, where we don't want to fall back to plain encoding.

Reply:
Added a comment.